### PR TITLE
[WIP] Reviewing OptimalBranchingCore

### PR DESCRIPTION
**Pull request by GiggleLiu**

### CODE REVIEW

**Comment by GiggleLiu** [comment]
@ArrogantGao What is NoProblem for?

**Comment by ArrogantGao** [comment]
> @ArrogantGao What is NoProblem for? Why do we need that? If you just want to define an interface, please checkout this example: https://github.com/GiggleLiu/ProblemReductions.jl/blob/main/src/models/models.jl
> 
> Also, we prefer writing documentation for the same function only once.

`NoProblem` provides a unified interface for all problem, which means the branching process is done, see 
https://github.com/ArrogantGao/OptimalBranching.jl/blob/5e2a9457a086e23afee099e09f0744018916acb1/lib/OptimalBranchingCore/src/branch.jl#L70

I agree to change the interface as that of the `ProblemReductions.jl`.

Another question is that do you think using problems type of `ProblemReductions.jl` in this package is a good idea?

File: docs/src/index.md [hunk 1]
```diff
@@ -45,20 +45,20 @@ julia> problem = MISProblem(g)
 MISProblem(20)
 
 # select the branching strategy
-julia> branching_strategy = OptBranchingStrategy(TensorNetworkSolver(), IPSolver(), EnvFilter(), MinBoundarySelector(2), D3Measure())
-OptBranchingStrategy
+julia> branching_strategy = BranchingStrategy(TensorNetworkSolver(), IPSolver(), EnvFilter(), MinBoundarySelector(2), D3Measure())
```
Comment (ArrogantGao) [inline docs/src/index.md:48]: We should allow user to define other branching strategies, not only the optimal branching strategy.

File: docs/src/index.md [hunk 2]
```diff
@@ -45,20 +45,20 @@ julia> problem = MISProblem(g)
...
+BranchingStrategy
     ├── table_solver - TensorNetworkSolver()
     ├── set_cover_solver - IPSolver(10)
     ├── pruner - EnvFilter()
     ├── selector - MinBoundarySelector(2)
     └── measure - D3Measure()
 
 
-julia> config = SolverConfig(MISReducer(), branching_strategy, MISSize)
+julia> config = SolverConfig(MISReducer(), branching_strategy, Int)
```
Comment (ArrogantGao) [inline docs/src/index.md:57]: I think `MISSize` is more clear, how should someone know what is represented by `Int`?

File: lib/OptimalBranchingMIS/src/reducer.jl [hunk 3]
```diff
@@ -45,30 +45,32 @@ function OptimalBranchingCore.problem_reduce(p::MISProblem, ::MISReducer, TR::Ty
...
 
 struct XiaoReducer <: AbstractReducer end
 
-function OptimalBranchingCore.problem_reduce(p::MISProblem, ::XiaoReducer, TR::Type{R}) where R<:AbstractResult
+function OptimalBranchingCore.reduce_problem(::Type{R}, p::MISProblem, ::XiaoReducer, ) where R
     g = p.g
     if nv(g) == 0
-        return [Branch(NoProblem(), 0)]
+        # NOTE: using NoProblem can be slower due to type instability.
+        return MISProblem(SimpleGraph(0)), R(0)
```
Comment (ArrogantGao) [inline lib/OptimalBranchingMIS/src/reducer.jl:69]: generate R(0) here is not correct, this gives MISCount(0, 1) and leads to incorrect count.

**Pull request merged**